File: backend/app/api/v1/analytics.py
"""Manager analytics.

⛔ ABSOLUTE PROHIBITION — READ BEFORE ADDING ANYTHING HERE

There must be no endpoint, query or response field that aggregates, ranks or
scores by user_id. No per-rep leaderboard, no rep performance metric, no
GROUP BY user_id. Aggregation is permitted at STORE and AREA level only.

Reason: if the score is tied to an individual, reps will photograph only
flattering angles and the entire dataset becomes worthless. This is a
labour-rights constraint agreed at design stage, not a missing feature. A
ticket asking for per-rep scoring should be escalated, not implemented.

`tests/integration/test_no_per_rep_analytics.py` enforces this by inspecting
this module's source.
"""

from __future__ import annotations

from datetime import UTC, datetime, timedelta
from uuid import UUID

from fastapi import APIRouter, Depends, Query
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.v1.deps import get_db, require_roles
from app.db.models import (
    Capture,
    GapFindingRow,
    ShelfAnalysisRow,
    Store,
    User,
    Visit,
)
from app.domain.enums import Role
from app.repositories import store_risk

router = APIRouter(tags=["analytics"])

# Analytics is management reporting, not part of a rep's job. Guarding at the
# router keeps a future endpoint from being added without a role check.
manager_only = require_roles(Role.MANAGER, Role.ADMIN, Role.DATA)


@router.get("/analytics/osa")
async def osa_trend(
    scope: str = Query("area", pattern="^(area|store)$"),
    area_id: str | None = Query(None, alias="areaId"),
    store_id: UUID | None = Query(None, alias="storeId"),
    days: int = Query(84, ge=1, le=365),
    db: AsyncSession = Depends(get_db),
    user: User = Depends(manager_only),
) -> dict:
    """OSA over time, bucketed by week. Store or area scope only."""
    since = datetime.now(UTC) - timedelta(days=days)

    query = (
        select(
            func.date_trunc("week", ShelfAnalysisRow.computed_at).label("bucket"),
            func.avg(ShelfAnalysisRow.osa_score).label("osa"),
            func.count(func.distinct(Visit.id)).label("visits"),
        )
        .join(Capture, Capture.id == ShelfAnalysisRow.capture_id)
        .join(Visit, Visit.id == Capture.visit_id)
        .join(Store, Store.id == Visit.store_id)
        .where(ShelfAnalysisRow.computed_at >= since)
        .group_by("bucket")
        .order_by("bucket")
    )
    if scope == "store" and store_id:
        query = query.where(Store.id == store_id)
    elif area_id:
        query = query.where(Store.area_id == area_id)

    rows = (await db.execute(query)).all()
    return {
        "scope": scope,
        "areaId": area_id,
        "storeId": str(store_id) if store_id else None,
        "points": [
            {
                "bucket": bucket.date().isoformat(),
                "osa": round(float(osa), 4),
                "visits": visits,
            }
            for bucket, osa, visits in rows
        ],
    }


@router.get("/analytics/risk-ranking")
async def risk_ranking(
    area_id: str | None = Query(None, alias="areaId"),
    limit: int = Query(20, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
    user: User = Depends(manager_only),
) -> dict:
    """Stores ranked by risk. Stores — never people."""
    query = select(Store)
    if area_id:
        query = query.where(Store.area_id == area_id)
    stores = (await db.execute(query)).scalars().all()

    # Shared with /v1/stores and /v1/routes/today. Three copies of this query
    # meant three chances for the rep's route and the manager's ranking to
    # disagree about which store is worst.
    facts = await store_risk.load(db, [s.id for s in stores])

    rows = []
    for store in stores:
        risk = store_risk.for_store(facts, store.id)
        rows.append(
            {
                "storeId": str(store.id),
                "storeName": store.name,
                "chain": store.chain,
                "storeFormat": store.store_format,
                "areaId": store.area_id,
                "lastOsa": risk.last_osa,
                "daysSinceLastVisit": risk.days_since_last_visit,
                "repeatGapSkus": risk.repeat_gap_skus,
                "riskScore": risk.score,
                "riskBand": risk.band,
            }
        )

    rows.sort(key=lambda r: -r["riskScore"])
    return {"areaId": area_id, "rows": rows[:limit]}


@router.get("/stores/{store_id}/history")
async def store_history(
    store_id: UUID,
    limit: int = Query(30, ge=1, le=200),
    db: AsyncSession = Depends(get_db),
    user: User = Depends(manager_only),
) -> dict:
    """Visit timeline for one store, with before/after OSA per visit."""
    visits = (
        (
            await db.execute(
                select(Visit)
                .where(Visit.store_id == store_id)
                .order_by(Visit.checked_in_at.desc())
                .limit(limit)
            )
        )
        .scalars()
        .all()
    )

    gap_counts = dict(
        (
            await db.execute(
                select(Capture.visit_id, func.count(GapFindingRow.id))
                .join(GapFindingRow, GapFindingRow.capture_id == Capture.id)
                .where(Capture.visit_id.in_([v.id for v in visits]) if visits else False)
                .group_by(Capture.visit_id)
            )
        ).all()
    )

    return {
        "storeId": str(store_id),
        "visits": [
            {
                "visitId": str(v.id),
                "checkedInAt": v.checked_in_at.isoformat(),
                "checkedOutAt": v.checked_out_at.isoformat() if v.checked_out_at else None,
                "osaBefore": float(v.osa_before) if v.osa_before is not None else None,
                "osaAfter": float(v.osa_after) if v.osa_after is not None else None,
                "gapsFound": gap_counts.get(v.id, 0),
                "gpsMatch": v.gps_match,
                "status": v.status,
            }
            for v in visits
        ],
    }

File: frontend/src/lib/api/analytics.ts
import { request, toPercent } from "@/lib/api/client";
import type { RiskBand } from "@/types";

/* Manager reporting.
 *
 * ⛔ Nothing in this file may hold, request or display a per-individual
 * figure. Aggregation is at store and area level only — a labour-rights
 * constraint agreed at design stage, enforced by tests on the backend. */

export interface Kpi {
  id: string;
  label: string;
  value: number;
  unit: string;
  /** change against the preceding window of equal length; null when there is
   *  no earlier window — which is not the same as "no change" */
  delta: number | null;
  deltaLabel: string;
  /** which direction counts as improvement */
  good: "up" | "down";
  /** null when the metric has no agreed target */
  target: number | null;
}

/** The headline cards.
 *
 *  The API returns only the KPIs it can compute, so the caller renders the
 *  array as it arrives. There is no fixed set of four: a card with no data
 *  behind it is absent, not zero and not estimated. */
export async function fetchKpis(areaId?: string, days = 28): Promise<Kpi[]> {
  const query = new URLSearchParams({ days: String(days) });
  if (areaId) query.set("areaId", areaId);
  const payload = await request<{ kpis: Kpi[] }>(`/v1/analytics/kpis?${query}`);
  return payload.kpis;
}

export interface Area {
  id: string;
  name: string;
  storeCount: number;
}

export function fetchAreas(): Promise<Area[]> {
  return request<Area[]>("/v1/areas");
}

export interface OsaPoint {
  /** ISO date of the week bucket */
  bucket: string;
  /** percent 0-100 */
  osa: number;
  visits: number;
}

export async function fetchOsaTrend(options: {
  areaId?: string;
  storeId?: string;
  days?: number;
} = {}): Promise<OsaPoint[]> {
  const query = new URLSearchParams({ days: String(options.days ?? 84) });
  if (options.storeId) {
    query.set("scope", "store");
    query.set("storeId", options.storeId);
  } else if (options.areaId) {
    query.set("areaId", options.areaId);
  }
  const payload = await request<{ points: { bucket: string; osa: number; visits: number }[] }>(
    `/v1/analytics/osa?${query}`,
  );
  return payload.points.map((p) => ({ ...p, osa: toPercent(p.osa, 1) ?? 0 }));
}

export interface RiskRow {
  storeId: string;
  storeName: string;
  chain: string;
  storeFormat: string;
  areaId: string;
  lastOsa: number | null;
  daysSinceLastVisit: number | null;
  repeatGapSkus: number;
  riskScore: number;
  riskBand: RiskBand;
}

export async function fetchRiskRanking(areaId?: string, limit = 20): Promise<RiskRow[]> {
  const query = new URLSearchParams({ limit: String(limit) });
  if (areaId) query.set("areaId", areaId);
  const payload = await request<{ rows: RiskRow[] }>(`/v1/analytics/risk-ranking?${query}`);
  return payload.rows.map((r) => ({
    ...r,
    lastOsa: toPercent(r.lastOsa),
    riskScore: toPercent(r.riskScore) ?? 0,
  }));
}

export interface PlannedStop extends Omit<RiskRow, "storeFormat"> {
  /** measured from this store's closed visits; null when it has none */
  avgVisitMinutes: number | null;
}

export async function fetchRoutePlan(areaId?: string): Promise<PlannedStop[]> {
  const query = areaId ? `?areaId=${encodeURIComponent(areaId)}` : "";
  const payload = await request<{ stops: PlannedStop[] }>(`/v1/analytics/route-plan${query}`);
  return payload.stops.map((s) => ({
    ...s,
    lastOsa: toPercent(s.lastOsa),
    riskScore: toPercent(s.riskScore) ?? 0,
  }));
}

export interface VisitCapture {
  captureId: string;
  category: string;
  shelfBayLabel: string;
  phase: "BEFORE" | "AFTER";
  capturedAt: string | null;
  /** percent 0-100, or null when the analysis has not landed yet */
  osaScore: number | null;
  modelVersion: string | null;
}

export interface StoreVisit {
  visitId: string;
  checkedInAt: string;
  checkedOutAt: string | null;
  osaBefore: number | null;
  osaAfter: number | null;
  gapsFound: number;
  gapsFixed: number;
  gpsMatch: boolean;
  status: string;
  /** the photographs behind this row — every number must be openable */
  captures: VisitCapture[];
}

/** One store's visit timeline. ⛔ Carries no identity — who visited is not
 *  part of what a manager reviews. */
export async function fetchStoreHistory(storeId: string, limit = 30): Promise<StoreVisit[]> {
  const payload = await request<{ visits: StoreVisit[] }>(
    `/v1/stores/${encodeURIComponent(storeId)}/history?limit=${limit}`,
  );
  return payload.visits.map((v) => ({
    ...v,
    osaBefore: toPercent(v.osaBefore),
    osaAfter: toPercent(v.osaAfter),
    captures: v.captures.map((c) => ({ ...c, osaScore: toPercent(c.osaScore) })),
  }));
}
